File: src/lib/api/incidents.ts
import { api } from "./client";

export interface Incident {
  id: string;
  app_id: string | null;
  app_name: string;
  title: string;
  severity: string;
  status: string;
  duration: string;
  assignee: string;
  ai_cause: string;
  health_impact: string;
  affected_deps: string[];
  timeline: Array<{ time: string; event: string; type: string }>;
  started_at: string;
  resolved_at: string;
}

export interface Alert {
  id: string;
  app_id: string | null;
  app_name: string;
  rule_name: string;
  metric: string;
  value: string;
  threshold: string;
  severity: string;
  status: string;
  fired_at: string;
  environment: string;
}

export const listIncidents = () => api.get<Incident[]>("/api/incidents");
export const listIncidentsByStatus = (status: string) => api.get<Incident[]>(`/api/incidents?status=${status}`);
export const getIncident = (id: string) => api.get<Incident>(`/api/incidents/${id}`);
export const updateIncident = (id: string, body: Partial<Incident>) => api.put<Incident>(`/api/incidents/${id}`, body);
export const acknowledgeIncident = (id: string) => api.put<Incident>(`/api/incidents/${id}`, { status: "acknowledged" });
export const resolveIncident = (id: string) => api.put<Incident>(`/api/incidents/${id}`, { status: "resolved" });
export const listAlerts = () => api.get<Alert[]>("/api/alerts");
export const listAlertsByStatus = (status: string) => api.get<Alert[]>(`/api/alerts?status=${status}`);
export const acknowledgeAlert = (id: string) => api.put<Alert>(`/api/alerts/${id}`, { status: "acknowledged" });

File: backend/app/models/incidents.py
from sqlalchemy import String, Integer, Float, Boolean, Text, ForeignKey, JSON
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database.session import Base
from app.models.base import TimestampMixin


class Incident(Base, TimestampMixin):
    __tablename__ = "incidents"

    id: Mapped[str] = mapped_column(String(64), primary_key=True)
    app_id: Mapped[str] = mapped_column(String(64), ForeignKey("applications.id"), nullable=True)
    title: Mapped[str] = mapped_column(String(256))
    severity: Mapped[str] = mapped_column(String(32), default="warning")
    status: Mapped[str] = mapped_column(String(32), default="active")
    duration: Mapped[str] = mapped_column(String(64), default="")
    assignee: Mapped[str] = mapped_column(String(128), default="")
    ai_cause: Mapped[str] = mapped_column(Text, default="")
    health_impact: Mapped[str] = mapped_column(String(64), default="")
    affected_deps: Mapped[str] = mapped_column(JSON, default=list)
    timeline: Mapped[str] = mapped_column(JSON, default=list)
    started_at: Mapped[str] = mapped_column(String(64), default="")
    resolved_at: Mapped[str] = mapped_column(String(64), default="")
    app_name: Mapped[str] = mapped_column(String(128), default="")

    app: Mapped["Application"] = relationship("Application", back_populates="incidents")

    from app.models.apps import Application


class Alert(Base, TimestampMixin):
    __tablename__ = "alerts"

    id: Mapped[str] = mapped_column(String(64), primary_key=True)
    app_id: Mapped[str] = mapped_column(String(64), ForeignKey("applications.id"), nullable=True)
    app_name: Mapped[str] = mapped_column(String(128), default="")
    rule_name: Mapped[str] = mapped_column(String(128))
    metric: Mapped[str] = mapped_column(String(64))
    value: Mapped[str] = mapped_column(String(64))
    threshold: Mapped[str] = mapped_column(String(64))
    severity: Mapped[str] = mapped_column(String(32), default="warning")
    status: Mapped[str] = mapped_column(String(32), default="firing")
    fired_at: Mapped[str] = mapped_column(String(64), default="")
    environment: Mapped[str] = mapped_column(String(64), default="Production")
